File: Information_Theory/GUI/labs/lab_1/task_2.py
import math
from collections import Counter

def calculate_entropy_and_histogram(text):
    # Преобразование текста в битовую последовательность
    bit_sequence = ''.join(format(ord(char), '08b') for char in text)

    # Подсчет частоты битов
    frequency = Counter(bit_sequence)
    total_bits = len(bit_sequence)

    # Вычисление энтропии текста
    entropy = 0
    for bit, freq in frequency.items():
        probability = freq / total_bits
        entropy += -probability * math.log2(probability)

    return entropy, frequency

def calculate_file_entropy(file_path):
    # Чтение содержимого файла в битовую последовательность
    with open(file_path, 'rb') as file:
        bit_sequence = ''.join(format(byte, '08b') for byte in file.read())

    # Подсчет частоты битов
    frequency = Counter(bit_sequence)
    total_bits = len(bit_sequence)

    # Вычисление энтропии файла
    entropy = 0
    for bit, freq in frequency.items():
        probability = freq / total_bits
        entropy += -probability * math.log2(probability)

    return entropy

File: Information_Theory/backend/labs/lab1/task2.js
const fs = require('fs')
const mammoth = require('mammoth')
const pdfjsLib = require('pdfjs-dist')

async function calculateTextEntropy(filePath) {
  let text

  // Определение типа файла и чтение текста
  if (filePath.endsWith('.txt')) {
    text = fs.readFileSync(filePath, 'utf-8')
  } else if (filePath.endsWith('.docx')) {
    const result = await mammoth.extractRawText({ path: filePath })
    text = result.value
  } else if (filePath.endsWith('.pdf')) {
    const data = new Uint8Array(fs.readFileSync(filePath))
    const pdf = await pdfjsLib.getDocument(data).promise
    let pdfText = ''
    for (let i = 1; i <= pdf.numPages; i++) {
      const page = await pdf.getPage(i)
      const content = await page.getTextContent()
      pdfText += content.items.map(item => item.str).join('')
    }
    text = pdfText
  } else {
    throw new Error('Unsupported file type')
  }

  // Вычисление энтропии текста
  const frequency = {}
  for (let i = 0; i < text.length; i++) {
    const char = text[i]
    if (frequency[char]) {
      frequency[char]++
    } else {
      frequency[char] = 1
    }
  }

  let entropy = 0
  const totalChars = text.length
  for (const char in frequency) {
    const probability = frequency[char] / totalChars
    entropy -= probability * Math.log2(probability)
  }

  // Возврат энтропии и имени каждого символа в тексте
  return { entropy, characters: Object.keys(frequency) }
}

module.exports = {
  calculateTextEntropy
}
